src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class AbstractFlowPass<TLocalState, TLocalFunctionState>
+    {
+        protected sealed class PendingBranchesCollection : IEnumerable<PendingBranch>
+        {
+            private ArrayBuilder<PendingBranch> _branches;
+            private PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>? _branchesToLabel;
+
+            internal PendingBranchesCollection()
+            {
+                _branches = ArrayBuilder<PendingBranch>.GetInstance();
+            }
+
+            internal void Free()
+            {
+                _branches.Free();
+                _branches = null!;
+                FreeBranchesToLabel();
+            }
+
+            internal void Clear()
+            {
+                _branches.Clear();
+                FreeBranchesToLabel();
+            }
+
+            private void FreeBranchesToLabel()
+            {
+                if (_branchesToLabel is { })
+                {
+                    foreach (var branches in _branchesToLabel.Values)
+                    {
+                        branches.Free();
+                    }
+                    _branchesToLabel.Free();
+                    _branchesToLabel = null;
+                }
+            }
+
+            internal ImmutableArray<PendingBranch> ToImmutable()
+            {
+                return _branchesToLabel is null ?
+                    _branches.ToImmutable() :
+                    ImmutableArray.CreateRange(this);
```

Comment:
>ImmutableArray.CreateRange(this); [](http://example.com/codeflow?start=20&length=33)

Would it be worth avoiding going through the IEnumerable?

Reply:
Had this thought as well. Digging in though there are a few places where we `foreach` the items here. Didn't see an obvious way to put into `for` loop. If we get rid of `IEnumerable` think we'd have to expand the reach of the change to account for that.

Reply:
I think the overhead of the `IEnumerable<T>` will be outweighed by the overhead of the `ImmutableArray<T>.Builder`.